File: 2017/PhysicsBasedAnimation/SpringAnimationBottomSheet/app/src/main/java/me/mehdi/springanimationbottomsheet/MainActivity.java
package me.mehdi.springanimationbottomsheet;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.GestureDetectorCompat;

import android.os.Bundle;
import android.view.MotionEvent;
import android.view.View;

public class MainActivity extends AppCompatActivity implements View.OnTouchListener {

    View mBottomSheet;
    GestureDetectorCompat mGestureDetector;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mBottomSheet = findViewById(R.id.bottomSheet);
        mGestureDetector = new GestureDetectorCompat(this, new CustomGestureListener(mBottomSheet));

        mBottomSheet.setOnTouchListener(this);
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        return mGestureDetector.onTouchEvent(event);
    }
}

File: 2022/MyPhysicalAnimationApp/app/src/main/java/com/codewithmehdi/myphysicalanimationapp/MainActivity.kt
package com.codewithmehdi.myphysicalanimationapp

import android.os.Bundle
import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.dynamicanimation.animation.DynamicAnimation
import androidx.dynamicanimation.animation.FlingAnimation
import com.codewithmehdi.myphysicalanimationapp.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity(){

    private lateinit var binding: ActivityMainBinding
    private var gestureDetectors: MutableMap<View, GestureDetector> = mutableMapOf()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)



        binding = ActivityMainBinding.inflate(layoutInflater)

        setContentView(binding.root)





        val myViews = listOf(
            binding.sun,
            binding.cloud1,
            binding.cloud2,
            binding.bird
        )

        myViews.forEach { image->
            gestureDetectors[image] = GestureDetector(this, MyOnGestureListener(image))
        }

        //Initialize gesture listeners
        myViews.forEach {image->
            image.setOnTouchListener { v, event ->
                gestureDetectors[v]!!.onTouchEvent(event)
            }
        }



    }
}



class MyOnGestureListener(private val view: View): SimpleOnGestureListener(){
    override fun onDown(e: MotionEvent?): Boolean {
        return true
    }

    override fun onFling(
        e1: MotionEvent?,
        e2: MotionEvent?,
        velocityX: Float,
        velocityY: Float
    ): Boolean {
        FlingAnimation(view, DynamicAnimation.Y).apply{
            setMinValue(0f)
            setMaxValue(view.resources.displayMetrics.heightPixels.toFloat() - view.height)
            setStartVelocity(velocityY)
            start()
        }

        FlingAnimation(view, DynamicAnimation.X).apply{
            setMinValue(0f)
            setMaxValue(view.resources.displayMetrics.widthPixels.toFloat() - view.width)
            setStartVelocity(velocityX)
            start()
        }

        return super.onFling(e1, e2, velocityX, velocityY)
    }
}